### PR TITLE
Add named parameters support

### PR DESCRIPTION
**Pull request by adoy**
This is a first iteration to support named arguments mocking. This branch do not support variadic function with named arguments.

### CODE REVIEW

**Comment by adoy** [comment]
@mlively This is an implementation to add named argument support (with variadic support). I had to change some logic but it should not break anything.

The approach I used is I added a `NamedArgumentResolver` that converts named arguments to positional arguments using Reflection (currently not cached but we may in the future add some reflection caching). If the function uses variadic params without name, params are checked as before using the chained matcher. If it's a named argument, then all parameters are checked at the same time.

I added some tests in the `tests/PhakeNamedArgumentsTest.php` file. I would really appreciate if you can have a look at the code to see if you see any edge cases that I did not think about, or any issue that this change may introduce.

File: tests/PhakeNamedArgumentsTest.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
+<?php
```
Comment (mlively) [inline tests/PhakeNamedArgumentsTest.php:1]: I don't think I have ever seen this before. Clever.

Comment (adoy) [reply tests/PhakeNamedArgumentsTest.php:1]: There is probably a better way to do it using Phpunit feature ? But this was the first thing that came to my mind at that point

**Review by mlively — APPROVED** [review]
How does this work with Phake::ignoreRemaining()? [documentation](https://phake.readthedocs.io/en/2.0/method-parameter-matchers.html#default-and-variable-parameters)


While this isn't how ignoreRemaining currently works, what would be interesting is if you could specify a single named argument in a parameter, then regardless of the other parameter values present in the matcher, if all specified parameters in the when() or verify() call match, and there is an indicator to ignore unspecified parameters then everything else would match. This could wind up being super powerful. I'd be okay with it not working this way out of the gate since it would definitely be separate and new functionality, but would be a great update. I am not sure what the syntax here would look like:

`
//method sig
public function methodUnderTest($param1, $param2, $param3) {}

//stub
Phake::when($mock)->methodUnderTest(Phake::ignoreUnspecified(), param1: "testvalue");

//The following would match
$mock->methodUnderTest("testvalue", "foo", "bar");
`

Thoughts?

**Comment by adoy** [comment]
You're right a new iteration with something like Phake::ignoreUnspecified() would be a nice and handy addition.
